File: typescript/src/codesandbox/filesystem.ts
import type { CompletedRunResult } from "../index.js";
import type { SandboxClient } from "./sandbox-client.js";
import {
  CodeSandboxError,
  type FSStatResult,
  type ReaddirEntry,
  translateArkerError,
} from "./types.js";

const STAT_FMT = "%y|%s|%T@|%A@|%C@\\n";
const READDIR_FMT = "%f|%y\\n";

function shellQuote(value: string): string {
  return `'${value.replace(/'/g, "'\\''")}'`;
}

function decode(bytes: Uint8Array): string {
  return new TextDecoder("utf-8", { fatal: false }).decode(bytes);
}

function kindToType(kind: string): "file" | "directory" {
  // Symlinks: resolved-kind best-effort approximation; the `isSymlink`
  // flag carries the real bit (matches @codesandbox/sdk).
  if (kind === "d") return "directory";
  return "file";
}

function isSymlinkKind(kind: string): boolean {
  return kind === "l";
}

/** Mirrors `@codesandbox/sdk` SandboxClient.fs. */
export class Filesystem {
  private readonly client: SandboxClient;

  constructor(client: SandboxClient) {
    this.client = client;
  }

  // ---- Native (Arker sync API) ----

  async readFile(path: string): Promise<Uint8Array> {
    try {
      return await this.client._sandbox._computer.sync.readFile(path);
    } catch (error) {
      throw translateArkerError(error);
    }
  }

  async readTextFile(path: string): Promise<string> {
    return decode(await this.readFile(path));
  }

  async writeFile(path: string, content: Uint8Array, opts: { create?: boolean; overwrite?: boolean } = {}): Promise<void> {
    await this.applyWriteOpts(path, opts);
    try {
      await this.client._sandbox._computer.sync.writeFile(path, content);
    } catch (error) {
      throw translateArkerError(error);
    }
  }

  async writeTextFile(path: string, content: string, opts: { create?: boolean; overwrite?: boolean } = {}): Promise<void> {
    await this.applyWriteOpts(path, opts);
    try {
      await this.client._sandbox._computer.sync.writeFile(path, content);
    } catch (error) {
      throw translateArkerError(error);
    }
  }

  private async applyWriteOpts(path: string, opts: { create?: boolean; overwrite?: boolean }): Promise<void> {
    const create = opts.create ?? true;
    const overwrite = opts.overwrite ?? true;
    if (create && overwrite) return;
    const { stdout } = await this.shell(`test -e ${shellQuote(path)}; echo $?`);
    const exists = stdout.trim() === "0";
    if (exists && !overwrite) {
      throw new CodeSandboxError(`writeFile(${path}): file exists and overwrite=false`);
    }
    if (!exists && !create) {
      throw new CodeSandboxError(`writeFile(${path}): file does not exist and create=false`);
    }
  }

  // ---- Shell-shim ----

  async readdir(path: string): Promise<ReaddirEntry[]> {
    const { stdout, exitCode } = await this.shell(
      `find ${shellQuote(path)} -maxdepth 1 -mindepth 1 -printf ${shellQuote(READDIR_FMT)}`,
    );
    if (exitCode !== 0) throw new CodeSandboxError(`readdir(${path}) failed`);
    const entries: ReaddirEntry[] = [];
    for (const line of stdout.split("\n")) {
      if (!line || !line.includes("|")) continue;
      const lastPipe = line.lastIndexOf("|");
      const name = line.slice(0, lastPipe);
      const kind = line.slice(lastPipe + 1);
      entries.push({ name, type: kindToType(kind), isSymlink: isSymlinkKind(kind) });
    }
    return entries;
  }

  async stat(path: string): Promise<FSStatResult> {
    const { stdout, exitCode } = await this.shell(
      `find ${shellQuote(path)} -maxdepth 0 -printf ${shellQuote(STAT_FMT)}`,
    );
    if (exitCode !== 0 || !stdout.trim()) {
      throw new CodeSandboxError(`stat(${path}) failed: not found`);
    }
    const parts = stdout.split("\n")[0]!.split("|");
    if (parts.length < 5) throw new CodeSandboxError(`stat(${path}): unparseable`);
    const [kind, sizeStr, mtimeStr, atimeStr, ctimeStr] = parts;
    return {
      type: kindToType(kind ?? ""),
      size: Number(sizeStr) || 0,
      atime: Number(atimeStr) || 0,
      mtime: Number(mtimeStr) || 0,
      ctime: Number(ctimeStr) || 0,
      isSymlink: isSymlinkKind(kind ?? ""),
    };
  }

  async mkdir(path: string, recursive: boolean = false): Promise<void> {
    const flag = recursive ? "-p" : "";
    const { stderr, exitCode } = await this.shell(`mkdir ${flag} ${shellQuote(path)}`.trim());
    if (exitCode !== 0) throw new CodeSandboxError(`mkdir(${path}) failed: ${stderr.trim()}`);
  }

  async remove(path: string, recursive: boolean = false): Promise<void> {
    const flag = recursive ? "-rf" : "-f";
    const { stderr, exitCode } = await this.shell(`rm ${flag} ${shellQuote(path)}`);
    if (exitCode !== 0) throw new CodeSandboxError(`remove(${path}) failed: ${stderr.trim()}`);
  }

  async rename(from: string, to: string, overwrite: boolean = false): Promise<void> {
    const parts = ["mv"];
    if (!overwrite) parts.push("-n");
    parts.push(shellQuote(from), shellQuote(to));
    const { stderr, exitCode } = await this.shell(parts.join(" "));
    if (exitCode !== 0) throw new CodeSandboxError(`rename failed: ${stderr.trim()}`);
  }

  async copy(from: string, to: string, recursive: boolean = false, overwrite: boolean = false): Promise<void> {
    const parts = ["cp"];
    if (recursive) parts.push("-r");
    if (!overwrite) parts.push("-n");
    parts.push(shellQuote(from), shellQuote(to));
    const { stderr, exitCode } = await this.shell(parts.join(" "));
    if (exitCode !== 0) throw new CodeSandboxError(`copy failed: ${stderr.trim()}`);
  }

  // ---- Not implemented ----

  watch(..._args: unknown[]): never {
    throw new Error("arker.codesandbox: fs.watch is not supported — no fs-event API.");
  }

  download(_path: string): never {
    throw new Error("arker.codesandbox: fs.download is not supported.");
  }

  batchWrite(..._args: unknown[]): never {
    throw new Error("arker.codesandbox: fs.batchWrite is not supported — loop over fs.writeFile.");
  }

  // ---- Internals ----

  private async shell(cmd: string): Promise<{ stdout: string; stderr: string; exitCode: number }> {
    let result;
    try {
      result = (await this.client._sandbox._computer.run(cmd)) as CompletedRunResult;
    } catch (error) {
      throw translateArkerError(error);
    }
    if (result.type !== "completed") {
      throw new CodeSandboxError(`unexpected run result type ${result.type}`);
    }
    return {
      stdout: decode(result.stdout),
      stderr: decode(result.stderr),
      exitCode: result.exitCode,
    };
  }
}

File: python/src/arker/codesandbox/_filesystem.py
"""`SandboxClient.fs` namespace.

Native ops (`readFile`/`writeFile`/`readTextFile`/`writeTextFile`) use Arker's
sync HTTP API. Shell-shim ops (`readdir`, `mkdir`, `remove`, `stat`, `rename`,
`copy`) go through `Computer.run`.
"""

from __future__ import annotations

import os
import shlex
from typing import TYPE_CHECKING

from ..computer import ArkerError, CompletedRunResult
from ._types import (
    CodeSandboxError,
    FSStatResult,
    ReaddirEntry,
    translate_arker_error,
)

if TYPE_CHECKING:
    from ._sandbox_client import SandboxClient


# `find ... -printf "%y|%s|%T@|%A@|%C@\n"` — kind, size, mtime, atime, ctime.
_STAT_FMT = "%y|%s|%T@|%A@|%C@\\n"

# `find ... -maxdepth 1 -mindepth 1 -printf "%f|%y\n"` for readdir.
_READDIR_FMT = "%f|%y\\n"


def _kind_to_type(kind: str) -> str:
    # Symlinks: `kind == "l"`. We don't resolve through to the target here;
    # codesandbox's daemon does. Best approximation: treat as "file" and let
    # the `is_symlink` flag carry the truth (matches their FileInfo shape).
    if kind == "d":
        return "directory"
    return "file"


def _is_symlink(kind: str) -> bool:
    return kind == "l"


class Filesystem:
    """Mirrors `@codesandbox/sdk` SandboxClient.fs."""

    def __init__(self, client: "SandboxClient") -> None:
        self._client = client

    # ---- Native (Arker sync API) ----

    def readFile(self, path: str) -> bytes:  # noqa: N802 — matches @codesandbox/sdk
        try:
            return self._client._sandbox._computer.sync.read_file(path)
        except ArkerError as error:
            raise translate_arker_error(error) from error

    def readTextFile(self, path: str) -> str:  # noqa: N802
        return self.readFile(path).decode("utf-8", errors="replace")

    def writeFile(self, path: str, content: bytes, opts: dict | None = None) -> None:  # noqa: N802
        self._apply_write_opts(path, opts)
        try:
            self._client._sandbox._computer.sync.write_file(path, bytes(content))
        except ArkerError as error:
            raise translate_arker_error(error) from error

    def writeTextFile(self, path: str, content: str, opts: dict | None = None) -> None:  # noqa: N802
        self._apply_write_opts(path, opts)
        try:
            self._client._sandbox._computer.sync.write_file(path, content)
        except ArkerError as error:
            raise translate_arker_error(error) from error

    def _apply_write_opts(self, path: str, opts: dict | None) -> None:
        """Honor codesandbox's `{create, overwrite}` semantics. Default for
        both is True (matches @codesandbox/sdk)."""
        if not opts:
            return
        create = opts.get("create", True)
        overwrite = opts.get("overwrite", True)
        if create and overwrite:
            return
        # Need to check existence to enforce either flag.
        exists_stdout, _, exists_code = self._shell(f"test -e {shlex.quote(path)}; echo $?")
        exists = exists_stdout.strip() == "0"
        if exists and not overwrite:
            raise CodeSandboxError(f"writeFile({path!r}): file exists and overwrite=False")
        if not exists and not create:
            raise CodeSandboxError(f"writeFile({path!r}): file does not exist and create=False")

    # Snake-case aliases for users coming from the Python conventions.
    read_file = readFile
    read_text_file = readTextFile
    write_file = writeFile
    write_text_file = writeTextFile

    # ---- Shell-shim ----

    def readdir(self, path: str) -> list[ReaddirEntry]:
        stdout, _, exit_code = self._shell(
            f"find {shlex.quote(path)} -maxdepth 1 -mindepth 1 -printf {shlex.quote(_READDIR_FMT)}"
        )
        if exit_code != 0:
            raise CodeSandboxError(f"readdir({path!r}) failed")
        entries: list[ReaddirEntry] = []
        for line in stdout.splitlines():
            if not line or "|" not in line:
                continue
            name, _, kind = line.rpartition("|")
            entries.append(ReaddirEntry(
                name=name,
                type=_kind_to_type(kind),
                is_symlink=_is_symlink(kind),
            ))
        return entries

    def stat(self, path: str) -> FSStatResult:
        stdout, _, exit_code = self._shell(
            f"find {shlex.quote(path)} -maxdepth 0 -printf {shlex.quote(_STAT_FMT)}"
        )
        if exit_code != 0 or not stdout.strip():
            raise CodeSandboxError(f"stat({path!r}) failed: not found")
        parts = stdout.splitlines()[0].split("|")
        if len(parts) < 5:
            raise CodeSandboxError(f"stat({path!r}): unparseable")
        kind, size_s, mtime_s, atime_s, ctime_s = parts[:5]
        return FSStatResult(
            type=_kind_to_type(kind),
            size=int(size_s) if size_s else 0,
            atime=float(atime_s) if atime_s else 0.0,
            mtime=float(mtime_s) if mtime_s else 0.0,
            ctime=float(ctime_s) if ctime_s else 0.0,
            is_symlink=_is_symlink(kind),
        )

    def mkdir(self, path: str, recursive: bool = False) -> None:
        flag = "-p" if recursive else ""
        _, stderr, exit_code = self._shell(f"mkdir {flag} {shlex.quote(path)}".strip())
        if exit_code != 0:
            raise CodeSandboxError(f"mkdir({path!r}) failed: {stderr.strip()}")

    def remove(self, path: str, recursive: bool = False) -> None:
        flag = "-rf" if recursive else "-f"
        _, stderr, exit_code = self._shell(f"rm {flag} {shlex.quote(path)}")
        if exit_code != 0:
            raise CodeSandboxError(f"remove({path!r}) failed: {stderr.strip()}")

    def rename(self, from_path: str, to_path: str, overwrite: bool = False) -> None:
        """Default `overwrite=False` matches `@codesandbox/sdk` — silent
        clobber would be a data-loss footgun."""
        parts = ["mv"]
        if not overwrite:
            parts.append("-n")
        parts.extend([shlex.quote(from_path), shlex.quote(to_path)])
        _, stderr, exit_code = self._shell(" ".join(parts))
        if exit_code != 0:
            raise CodeSandboxError(f"rename failed: {stderr.strip()}")

    def copy(self, from_path: str, to_path: str, recursive: bool = False, overwrite: bool = False) -> None:
        """Default `overwrite=False` matches `@codesandbox/sdk`."""
        parts = ["cp"]
        if recursive:
            parts.append("-r")
        if not overwrite:
            parts.append("-n")
        parts.extend([shlex.quote(from_path), shlex.quote(to_path)])
        _, stderr, exit_code = self._shell(" ".join(parts))
        if exit_code != 0:
            raise CodeSandboxError(f"copy failed: {stderr.strip()}")

    # ---- Not implemented ----

    def watch(self, *args, **kwargs):
        raise NotImplementedError(
            "arker.codesandbox: fs.watch is not supported — no fs-event API. "
            "Poll readdir/stat if needed."
        )

    def download(self, path: str) -> dict:
        raise NotImplementedError(
            "arker.codesandbox: fs.download is not supported — Arker has no "
            "signed download URL endpoint. Use fs.readFile() and write locally."
        )

    def batchWrite(self, *args, **kwargs):  # noqa: N802
        raise NotImplementedError(
            "arker.codesandbox: fs.batchWrite is not supported — loop over fs.writeFile."
        )

    # ---- Internals ----

    def _shell(self, cmd: str) -> tuple[str, str, int]:
        try:
            result = self._client._sandbox._computer.run(cmd)
        except ArkerError as error:
            raise translate_arker_error(error) from error
        if not isinstance(result, CompletedRunResult):
            raise CodeSandboxError(f"unexpected run result type {type(result).__name__}")
        return (
            result.stdout.decode("utf-8", errors="replace"),
            result.stderr.decode("utf-8", errors="replace"),
            result.exit_code,
        )
